osgp/platform/osgp-secret-management/src/main/java/org/opensmartgridplatform/secretmanagement/application/services/SecretManagementService.java

```diff
@@ -18,146 +23,285 @@
+        private EncryptedTypedSecret(byte[] secret, SecretType type, String encryptionKeyReference) {
+            if(secret==null) {
+                throw new IllegalArgumentException("No NULL value allowed for parameter 'secret'");
+            } else if(type==null) {
+                throw new IllegalArgumentException("No NULL value allowed for parameter 'secretType'");
+            }
+            this.encryptedSecret = secret;
+            this.type = type;
+            this.encryptionKeyReference = encryptionKeyReference;
+        }
+
+        private boolean hasNullSecret() {
+            return this.encryptedSecret==null;
+        }
+
+        private TypedSecret toTypedSecret() {
+            return new TypedSecret(this.encryptedSecret, this.type);
+        }
+
+        private static EncryptedTypedSecret getNullInstance(SecretType type) {
+            return new EncryptedTypedSecret(type);
+        }
+    }
+
     private final EncryptionDelegate encryptionDelegate;
     private final EncryptionProviderType encryptionProviderType;
     private final DbEncryptedSecretRepository secretRepository;
     private final DbEncryptionKeyRepository keyRepository;
+    private final RsaEncrypter rsaEncrypter;
 
     public SecretManagementService(
             @Qualifier("DefaultEncryptionDelegate") final EncryptionDelegate defaultEncryptionDelegate,
             final EncryptionProviderType encryptionProviderType, final DbEncryptedSecretRepository secretRepository,
-            final DbEncryptionKeyRepository keyRepository) {
+            final DbEncryptionKeyRepository keyRepository, RsaEncrypter rsaEncrypter) {
         this.encryptionDelegate = defaultEncryptionDelegate;
         this.encryptionProviderType = encryptionProviderType;
         this.secretRepository = secretRepository;
         this.keyRepository = keyRepository;
+        this.rsaEncrypter = rsaEncrypter;
     }
 
-    public void storeSecrets(final String deviceIdentification, final List<TypedSecret> secrets) {
-        //@formatter:off
-        secrets.stream()
-                .map(t -> this.validateSecret(deviceIdentification, t))
-                .map(s -> this.createEncrypted(deviceIdentification, s, this.getKey()))
-                .forEach(this.secretRepository::save);
-        //@formatter:on
-    }
-
-    private DbEncryptionKeyReference getKey() {
+    private DbEncryptionKeyReference getCurrentKey() {
         final Date now = new Date();
-        final Page<DbEncryptionKeyReference> keyRefsPage = this.keyRepository.findByTypeAndValid(now,
-                this.encryptionProviderType, Pageable.unpaged());
-        if (keyRefsPage.getSize() > 1) {
+        final List<DbEncryptionKeyReference> keyRefs = this.keyRepository
+                .findByTypeAndValid(this.encryptionProviderType, now);
+        if (keyRefs.size() > 1) {
             throw new IllegalStateException("Multiple encryption keys found that are valid at " + now);
+        } else if(keyRefs.size()==0) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Use isEmpty() to check whether the collection is empty or not. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1155)